jgrapht-guava/src/main/java/org/jgrapht/graph/guava/BaseGraphAdapter.java

```diff
@@ -0,0 +1,214 @@
+    extends AbstractGraph<V, EndpointPair<V>>
+    implements Graph<V, EndpointPair<V>>, Cloneable, Serializable
+{
+    private static final long serialVersionUID = -6742507788742087708L;
+
+    protected static final String LOOPS_NOT_ALLOWED = "loops not allowed";
+
+    protected transient Set<V> unmodifiableVertexSet = null;
+    protected transient Set<EndpointPair<V>> unmodifiableEdgeSet = null;
+
+    protected transient G graph;
+
+    /**
+     * Create a new adapter.
+     * 
+     * @param graph the graph
+     */
+    public BaseGraphAdapter(G graph)
+    {
+        this.graph = Objects.requireNonNull(graph);
+    }
+
+    @Override
+    public EndpointPair<V> getEdge(V sourceVertex, V targetVertex)
+    {
+        if (sourceVertex == null || targetVertex == null) {
+            return null;
+        } else if (!graph.hasEdgeConnecting(sourceVertex, targetVertex)) {
+            return null;
+        } else {
+            return createEdge(sourceVertex, targetVertex);
+        }
+    }
+
+    @Override
+    public EdgeFactory<V, EndpointPair<V>> getEdgeFactory()
+    {
+        return this::createEdge;
+    }
+
+    @Override
+    public Set<V> vertexSet()
+    {
+        if (unmodifiableVertexSet == null) {
+            unmodifiableVertexSet = Collections.unmodifiableSet(graph.nodes());
+        }
+        return unmodifiableVertexSet;
+    }
+
+    @Override
+    public V getEdgeSource(EndpointPair<V> e)
+    {
+        return e.nodeU();
+    }
+
+    @Override
+    public V getEdgeTarget(EndpointPair<V> e)
+    {
+        return e.nodeV();
+    }
+
+    @Override
+    public GraphType getType()
+    {
+        return (graph.isDirected() ? new DefaultGraphType.Builder().directed()
+            : new DefaultGraphType.Builder().undirected())
+                .weighted(false).allowMultipleEdges(false)
+                .allowSelfLoops(graph.allowsSelfLoops()).build();
+    }
+
+    @Override
+    public boolean containsEdge(EndpointPair<V> e)
+    {
+        return graph.edges().contains(e);
+    }
+
+    @Override
+    public boolean containsVertex(V v)
+    {
+        return graph.nodes().contains(v);
+    }
+
+    @Override
+    public Set<EndpointPair<V>> edgeSet()
+    {
+        if (unmodifiableEdgeSet == null) {
+            unmodifiableEdgeSet = Collections.unmodifiableSet(graph.edges());
+        }
+        return unmodifiableEdgeSet;
+    }
+
+    @Override
+    public int degreeOf(V vertex)
+    {
+        return graph.degree(vertex);
+    }
+
+    @Override
+    public Set<EndpointPair<V>> edgesOf(V vertex)
+    {
+        return graph.incidentEdges(vertex);
+    }
+
+    @Override
+    public int inDegreeOf(V vertex)
+    {
+        return graph.inDegree(vertex);
+    }
+
+    @Override
+    public Set<EndpointPair<V>> incomingEdgesOf(V vertex)
+    {
+        return graph
+            .predecessors(vertex).stream().map(other -> createEdge(other, vertex))
+            .collect(Collectors.toSet());
```

Comment:
Considering that:

1. The current Guava graph implementations generally return unmodifiable views for methods that return `Set`s (according to https://github.com/google/guava/wiki/GraphsExplained#accessor-behavior), and
2. `Collectors.toSet()` returns a mutable set as of Java 8 and that users may expect that if they modify this set, then the underlying graph will also be changed

...I think we should swap out this line of code for this one instead:

```java
    .collect(collectingAndThen(toSet(), Collections::unmodifiableSet));
```